consensus/istanbul/validator/default_test.go

```diff
@@ -211,74 +209,11 @@ func testStickyProposer(t *testing.T) {
-	if valSet.QuorumSize(IBFT_FORMULA_CEIL_2N_3) != 2 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", IBFT_FORMULA_CEIL_2N_3, 2, 2, valSet.QuorumSize(IBFT_FORMULA_CEIL_2N_3))
-	}
-	// formulaType = IBFT_FORMULA_FLOOR_2N_3_PLUS_1, proposed update floor(2N/3)+1
-	if valSet.QuorumSize(IBFT_FORMULA_FLOOR_2N_3_PLUS_1) != 2 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", IBFT_FORMULA_FLOOR_2N_3_PLUS_1, 2, 2, valSet.QuorumSize(IBFT_FORMULA_FLOOR_2N_3_PLUS_1))
-	}
-	// N==3
-	b3 := common.Hex2Bytes(testAddress3)
-	addr3 := common.BytesToAddress(b3)
-	valSet.AddValidator(addr3)
-	if valSet.Size() != 3 {
-		t.Errorf("valSet.Size() expected: %v, got: %v", 3, valSet.Size())
-	}
-	if valSet.F() != 0 {
-		t.Errorf("valSet.F() expected: %v, got: %v", 0, valSet.F())
-	}
-	// formulaType = 0, default 2f()+1 
-	if valSet.QuorumSize(0) != 1 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", 0, 3, 1, valSet.QuorumSize(0))
-	}
-	// formulaType = IBFT_FORMULA_CEIL_2N_3, proposed update Ceil(2N/3)
-	if valSet.QuorumSize(IBFT_FORMULA_CEIL_2N_3) != 2 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", IBFT_FORMULA_CEIL_2N_3, 3, 2, valSet.QuorumSize(IBFT_FORMULA_CEIL_2N_3))
-	}
-	// formulaType = IBFT_FORMULA_FLOOR_2N_3_PLUS_1, proposed update floor(2N/3)+1 
-	if valSet.QuorumSize(IBFT_FORMULA_FLOOR_2N_3_PLUS_1) != 3 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", IBFT_FORMULA_FLOOR_2N_3_PLUS_1, 3, 3, valSet.QuorumSize(IBFT_FORMULA_FLOOR_2N_3_PLUS_1))
-	}
-
-	// N==4
-	b4 := common.Hex2Bytes(testAddress4)
-	addr4 := common.BytesToAddress(b4)
-	valSet.AddValidator(addr4)
-	if valSet.Size() != 4 {
-		t.Errorf("valSet.Size() expected: %v, got: %v", 4, valSet.Size())
-	}
-	if valSet.F() != 1 {
-		t.Errorf("valSet.F() expected: %v, got: %v", 1, valSet.F())
-	}
-	// formulaType = 0, default 2f()+1 
-	if valSet.QuorumSize(0) != 3 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", 0, 4, 3, valSet.QuorumSize(0))
-	}
-	// formulaType = IBFT_FORMULA_CEIL_2N_3, proposed update Ceil(2N/3)
-	if valSet.QuorumSize(IBFT_FORMULA_CEIL_2N_3) != 3 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", IBFT_FORMULA_CEIL_2N_3, 4, 3, valSet.QuorumSize(IBFT_FORMULA_CEIL_2N_3))
-	}
-	// formulaType = IBFT_FORMULA_FLOOR_2N_3_PLUS_1, proposed update floor(2N/3)+1 
-	if valSet.QuorumSize(IBFT_FORMULA_FLOOR_2N_3_PLUS_1) != 3 {
-		t.Errorf("QuorumSize wrong for formulaType: %v,  N: %v, expected: %v, got: %v", IBFT_FORMULA_FLOOR_2N_3_PLUS_1, 4, 3, valSet.QuorumSize(IBFT_FORMULA_FLOOR_2N_3_PLUS_1))
+	valSet := newDefaultSet([]common.Address{}, istanbul.RoundRobin)
+	for i:=1; i<=1000; i++ {
+		valSet.AddValidator(common.StringToAddress(string(i)));
+		if valSet.QuorumSize() <= (valSet.Size() - valSet.QuorumSize() + valSet.F()) {
```

Comment:
You got the right idea, but let's make the condition test for a tight bound, not just a lower bound. To ensure this, we must also check that QuorumSize() is the lowest number for which the invariant is true, i.e., QuorumSize() - 1 > valSet.Size() - (QuorumSize() - 1) + f must also be true. We can simplify the expressions such that the test fails upon: 2 * valSet.QuorumSize() <= valSet.Size() + valSet.F() || 2 * valSet.QuorumSize() > valSet.Size() + valSet.F() + 2. Thanks!